conda/gateways/repodata/jlap/interface.py

```diff
@@ -34,45 +42,85 @@ def __init__(
+
+    def repodata_parsed(self, state: dict | RepodataState) -> dict | None:
+        """
+        JLAP has to parse the JSON anyway.
+
+        Use this to avoid a redundant parse when repodata is updated.
+
+        When repodata is not updated, it doesn't matter whether this function or
+        the caller reads from a file.
+        """
         session = CondaSession()
 
         repodata_url = f"{self._url}/{self._repodata_fn}"
-        # jlap_url = f"{self._url}/{self._repodata_fn}"[: -len(".json")] + ".jlap"
 
         # XXX won't modify caller's state dict
         state_ = RepodataState(dict=state)
 
-        def get_place(url, extra=""):
-            if url == repodata_url and extra == "":
-                return self._cache_path_json
-            raise NotImplementedError("Unexpected URL", url)
+        # at this point, self._cache.state == state == state_
 
+        temp_path = (
+            self._cache.cache_dir / f"{self._cache.name}.{os.urandom(2).hex()}.tmp"
+        )
         try:
             with conda_http_errors(self._url, self._repodata_fn):
-                fetch.request_url_jlap_state(
-                    repodata_url, state_, get_place=get_place, session=session
+                repodata_json_or_none = fetch.request_url_jlap_state(
+                    repodata_url,
+                    state_,
+                    session=session,
+                    cache=self._cache,
+                    temp_path=temp_path,
                 )
-        except fetch.Jlap304NotModified:
-            raise Response304ContentUnchanged()
 
-        # XXX update caller's state dict-or-RepodataState
-        state.update(state_)
+                # update caller's state dict-or-RepodataState. Do this before
+                # the self._cache.replace() call which also writes state, then
+                # signal not to write state to caller.
+                state.update(state_)
 
-        state["_url"] = self._url
-        headers = state.get("jlap", {}).get(
-            "headers"
-        )  # XXX overwrite headers in jlapper.request_url_jlap_state
-        if headers:
-            state["_etag"] = headers.get("etag")
-            state["_mod"] = headers.get("last-modified")
-            state["_cache_control"] = headers.get("cache-control")
+                state[URL_KEY] = self._url
+                headers = state.get("jlap", {}).get(
+                    "headers"
+                )  # XXX overwrite headers in jlapper.request_url_jlap_state
+                if headers:
+                    state[ETAG_KEY] = headers.get("etag")
+                    state[LAST_MODIFIED_KEY] = headers.get("last-modified")
+                    state[CACHE_CONTROL_KEY] = headers.get("cache-control")
 
-        # Indicate that subdir_data mustn't rewrite cache_path_json
-        raise RepodataOnDisk()
+                self._cache.state.update(state)
+
+            if temp_path.exists():
+                self._cache.replace(temp_path)
```

Comment:
Need to find an elegant way to prevent the caller from also calling `cache.refresh()` writing the (small) state/info file again.